File: core/src/main/scala/org/apache/spark/sql/delta/DeltaParquetFileFormat.scala
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.delta

import java.net.URI

import scala.collection.mutable.ArrayBuffer
import scala.util.control.NonFatal

import org.apache.spark.sql.delta.RowIndexFilterType
import org.apache.spark.sql.delta.DeltaParquetFileFormat._
import org.apache.spark.sql.delta.actions.{DeletionVectorDescriptor, Metadata, Protocol}
import org.apache.spark.sql.delta.deletionvectors.{DropMarkedRowsFilter, KeepAllRowsFilter, KeepMarkedRowsFilter}
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path

import org.apache.spark.broadcast.Broadcast
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.execution.datasources.PartitionedFile
import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
import org.apache.spark.sql.execution.vectorized.{OffHeapColumnVector, OnHeapColumnVector, WritableColumnVector}
import org.apache.spark.sql.internal.SQLConf
import org.apache.spark.sql.sources.Filter
import org.apache.spark.sql.types.{ByteType, LongType, StructField, StructType}
import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
import org.apache.spark.util.SerializableConfiguration

/**
 * A thin wrapper over the Parquet file format to support
 *  - columns names without restrictions.
 *  - populated a column from the deletion vector of this file (if exists) to indicate
 *    whether the row is deleted or not according to the deletion vector. Consumers
 *    of this scan can use the column values to filter out the deleted rows.
 */
case class DeltaParquetFileFormat(
    protocol: Protocol,
    metadata: Metadata,
    isSplittable: Boolean = true,
    disablePushDowns: Boolean = false,
    tablePath: Option[String] = None,
    broadcastDvMap: Option[Broadcast[Map[URI, DeletionVectorDescriptorWithFilterType]]] = None,
    broadcastHadoopConf: Option[Broadcast[SerializableConfiguration]] = None)
  extends ParquetFileFormat {
  // Validate either we have all arguments for DV enabled read or none of them.
  if (hasDeletionVectorMap) {
    require(tablePath.isDefined && !isSplittable && disablePushDowns,
      "Wrong arguments for Delta table scan with deletion vectors")
  }

  val columnMappingMode: DeltaColumnMappingMode = metadata.columnMappingMode
  val referenceSchema: StructType = metadata.schema

  if (columnMappingMode == IdMapping) {
    val requiredReadConf = SQLConf.PARQUET_FIELD_ID_READ_ENABLED
    require(SparkSession.getActiveSession.exists(_.sessionState.conf.getConf(requiredReadConf)),
      s"${requiredReadConf.key} must be enabled to support Delta id column mapping mode")
    val requiredWriteConf = SQLConf.PARQUET_FIELD_ID_WRITE_ENABLED
    require(SparkSession.getActiveSession.exists(_.sessionState.conf.getConf(requiredWriteConf)),
      s"${requiredWriteConf.key} must be enabled to support Delta id column mapping mode")
  }

  def prepareSchema(inputSchema: StructType): StructType = {
    DeltaColumnMapping.createPhysicalSchema(inputSchema, referenceSchema, columnMappingMode)
  }

  override def isSplitable(
    sparkSession: SparkSession, options: Map[String, String], path: Path): Boolean = isSplittable

  def hasDeletionVectorMap(): Boolean = broadcastDvMap.isDefined && broadcastHadoopConf.isDefined

  /**
   * We sometimes need to replace FileFormat within LogicalPlans, so we have to override
   * `equals` to ensure file format changes are captured
   */
  override def equals(other: Any): Boolean = {
    other match {
      case ff: DeltaParquetFileFormat =>
        ff.columnMappingMode == columnMappingMode &&
        ff.referenceSchema == referenceSchema &&
        ff.isSplittable == isSplittable &&
        ff.disablePushDowns == disablePushDowns
      case _ => false
    }
  }

  override def hashCode(): Int = getClass.getCanonicalName.hashCode()

  override def buildReaderWithPartitionValues(
      sparkSession: SparkSession,
      dataSchema: StructType,
      partitionSchema: StructType,
      requiredSchema: StructType,
      filters: Seq[Filter],
      options: Map[String, String],
      hadoopConf: Configuration): PartitionedFile => Iterator[InternalRow] = {
    val pushdownFilters = if (disablePushDowns) Seq.empty else filters

    val parquetDataReader: PartitionedFile => Iterator[InternalRow] =
      super.buildReaderWithPartitionValues(
        sparkSession,
        prepareSchema(dataSchema),
        prepareSchema(partitionSchema),
        prepareSchema(requiredSchema),
        pushdownFilters,
        options,
        hadoopConf)

    val schemaWithIndices = requiredSchema.fields.zipWithIndex
    def findColumn(name: String): Option[ColumnMetadata] = {
      val results = schemaWithIndices.filter(_._1.name == name)
      if (results.length > 1) {
        throw new IllegalArgumentException(
          s"There are more than one column with name=`$name` requested in the reader output")
      }
      results.headOption.map(e => ColumnMetadata(e._2, e._1))
    }
    val isRowDeletedColumn = findColumn(IS_ROW_DELETED_COLUMN_NAME)
    val rowIndexColumn = findColumn(ROW_INDEX_COLUMN_NAME)

    if (isRowDeletedColumn.isEmpty && rowIndexColumn.isEmpty) {
      return parquetDataReader // no additional metadata is needed.
    } else {
      // verify the file splitting and filter pushdown are disabled. The new additional
      // metadata columns cannot be generated with file splitting and filter pushdowns
      require(!isSplittable, "Cannot generate row index related metadata with file splitting")
      require(disablePushDowns, "Cannot generate row index related metadata with filter pushdown")
    }

    if (hasDeletionVectorMap && isRowDeletedColumn.isEmpty) {
        throw new IllegalArgumentException(
          s"Expected a column $IS_ROW_DELETED_COLUMN_NAME in the schema")
    }

    val useOffHeapBuffers = sparkSession.sessionState.conf.offHeapColumnVectorEnabled
    (partitionedFile: PartitionedFile) => {
      val rowIteratorFromParquet = parquetDataReader(partitionedFile)
      val iterToReturn =
        iteratorWithAdditionalMetadataColumns(
          partitionedFile,
          rowIteratorFromParquet,
          isRowDeletedColumn,
          useOffHeapBuffers = useOffHeapBuffers,
          rowIndexColumn = rowIndexColumn)
      iterToReturn.asInstanceOf[Iterator[InternalRow]]
    }
  }

  override def supportFieldName(name: String): Boolean = {
    if (columnMappingMode != NoMapping) true else super.supportFieldName(name)
  }

  def copyWithDVInfo(
      tablePath: String,
      broadcastDvMap: Broadcast[Map[URI, DeletionVectorDescriptorWithFilterType]],
      broadcastHadoopConf: Broadcast[SerializableConfiguration]): DeltaParquetFileFormat = {
    this.copy(
      isSplittable = false,
      disablePushDowns = true,
      tablePath = Some(tablePath),
      broadcastDvMap = Some(broadcastDvMap),
      broadcastHadoopConf = Some(broadcastHadoopConf))
  }

  /**
   * Modifies the data read from underlying Parquet reader by populating one or both of the
   * following metadata columns.
   *   - [[IS_ROW_DELETED_COLUMN_NAME]] - row deleted status from deletion vector corresponding
   *   to this file
   *   - [[ROW_INDEX_COLUMN_NAME]] - index of the row within the file.
   */
  private def iteratorWithAdditionalMetadataColumns(
      partitionedFile: PartitionedFile,
      iterator: Iterator[Object],
      isRowDeletedColumn: Option[ColumnMetadata],
      rowIndexColumn: Option[ColumnMetadata],
      useOffHeapBuffers: Boolean): Iterator[Object] = {
    val pathUri = new URI(partitionedFile.filePath)

    val rowIndexFilter = isRowDeletedColumn.map { col =>
      // Fetch the DV descriptor from the broadcast map and create a row index filter
      broadcastDvMap.get.value
        .get(pathUri)
        .map { case DeletionVectorDescriptorWithFilterType(dvDescriptor, filterType) =>
          filterType match {
            case i if i == RowIndexFilterType.IF_CONTAINED =>
              DropMarkedRowsFilter.createInstance(
                dvDescriptor,
                broadcastHadoopConf.get.value.value,
                tablePath.map(new Path(_)))
            case i if i == RowIndexFilterType.IF_NOT_CONTAINED =>
              KeepMarkedRowsFilter.createInstance(
                dvDescriptor,
                broadcastHadoopConf.get.value.value,
                tablePath.map(new Path(_)))
          }
        }
        .getOrElse(KeepAllRowsFilter)
    }

    val metadataColumns = Seq(isRowDeletedColumn, rowIndexColumn).filter(_.nonEmpty).map(_.get)

    // Unfortunately there is no way to verify the Parquet index is starting from 0.
    // We disable the splits, so the assumption is ParquetFileFormat respects that
    var rowIndex: Long = 0

    // Used only when non-column row batches are received from the Parquet reader
    val tempVector = new OnHeapColumnVector(1, ByteType)

    iterator.map { row =>
      row match {
        case batch: ColumnarBatch => // When vectorized Parquet reader is enabled
          val size = batch.numRows()
          // Create vectors for all needed metadata columns.
          // We can't use the one from Parquet reader as it set the
          // [[WritableColumnVector.isAllNulls]] to true and it can't be reset with using any
          // public APIs.
          trySafely(useOffHeapBuffers, size, metadataColumns) { writableVectors =>
            val indexVectorTuples = new ArrayBuffer[(Int, ColumnVector)]
            var index = 0
            isRowDeletedColumn.foreach { columnMetadata =>
              val isRowDeletedVector = writableVectors(index)
              rowIndexFilter.get
                .materializeIntoVector(rowIndex, rowIndex + size, isRowDeletedVector)
              indexVectorTuples += (columnMetadata.index -> isRowDeletedVector)
              index += 1
            }

            rowIndexColumn.foreach { columnMetadata =>
              val rowIndexVector = writableVectors(index)
              // populate the row index column value
              for (i <- 0 until size) {
                rowIndexVector.putLong(i, rowIndex + i)
              }

              indexVectorTuples += (columnMetadata.index -> rowIndexVector)
              index += 1
            }

            val newBatch = replaceVectors(batch, indexVectorTuples.toSeq: _*)
            rowIndex += size
            newBatch
          }

        case rest: InternalRow => // When vectorized Parquet reader is disabled
          // Temporary vector variable used to get DV values from RowIndexFilter
          // Currently the RowIndexFilter only supports writing into a columnar vector
          // and doesn't have methods to get DV value for a specific row index.
          // TODO: This is not efficient, but it is ok given the default reader is vectorized
          // reader and this will be temporary until Delta upgrades to Spark with Parquet
          // reader that automatically generates the row index column.
          isRowDeletedColumn.foreach { columnMetadata =>
            rowIndexFilter.get.materializeIntoVector(rowIndex, rowIndex + 1, tempVector)
            rest.setLong(columnMetadata.index, tempVector.getByte(0))
          }

          rowIndexColumn.foreach(columnMetadata => rest.setLong(columnMetadata.index, rowIndex))
          rowIndex += 1
          rest
        case others =>
          throw new RuntimeException(
            s"Parquet reader returned an unknown row type: ${others.getClass.getName}")
      }
    }
  }
}

object DeltaParquetFileFormat {
  /**
   * Column name used to identify whether the row read from the parquet file is marked
   * as deleted according to the Delta table deletion vectors
   */
  val IS_ROW_DELETED_COLUMN_NAME = "__delta_internal_is_row_deleted"
  val IS_ROW_DELETED_STRUCT_FIELD = StructField(IS_ROW_DELETED_COLUMN_NAME, ByteType)

  /** Row index for each column */
  val ROW_INDEX_COLUMN_NAME = "__delta_internal_row_index"
  val ROW_INDEX_STRUCT_FILED = StructField(ROW_INDEX_COLUMN_NAME, LongType)

  /** Utility method to create a new writable vector */
  private def newVector(
      useOffHeapBuffers: Boolean, size: Int, dataType: StructField): WritableColumnVector = {
    if (useOffHeapBuffers) {
      OffHeapColumnVector.allocateColumns(size, Seq(dataType).toArray)(0)
    } else {
      OnHeapColumnVector.allocateColumns(size, Seq(dataType).toArray)(0)
    }
  }

  /** Try the operation, if the operation fails release the created resource */
  private def trySafely[R <: WritableColumnVector, T](
      useOffHeapBuffers: Boolean,
      size: Int,
      columns: Seq[ColumnMetadata])(f: Seq[WritableColumnVector] => T): T = {
    val resources = new ArrayBuffer[WritableColumnVector](columns.size)
    try {
      columns.foreach(col => resources.append(newVector(useOffHeapBuffers, size, col.structField)))
      f(resources.toSeq)
    } catch {
      case NonFatal(e) =>
        resources.foreach(closeQuietly(_))
        throw e
    }
  }

  /** Utility method to quietly close an [[AutoCloseable]] */
  private def closeQuietly(closeable: AutoCloseable): Unit = {
    if (closeable != null) {
      try {
        closeable.close()
      } catch {
        case NonFatal(_) => // ignore
      }
    }
  }

  /**
   * Helper method to replace the vectors in given [[ColumnarBatch]].
   * New vectors and its index in the batch are given as tuples.
   */
  private def replaceVectors(
      batch: ColumnarBatch,
      indexVectorTuples: (Int, ColumnVector) *): ColumnarBatch = {
    val vectors = ArrayBuffer[ColumnVector]()
    for (i <- 0 until batch.numCols()) {
      var replaced: Boolean = false
      for (indexVectorTuple <- indexVectorTuples) {
        val index = indexVectorTuple._1
        val vector = indexVectorTuple._2
        if (indexVectorTuple._1 == i) {
          vectors += indexVectorTuple._2
          // Make sure to close the existing vector allocated in the Parquet
          batch.column(i).close()
          replaced = true
        }
      }
      if (!replaced) {
        vectors += batch.column(i)
      }
    }
    new ColumnarBatch(vectors.toArray, batch.numRows())
  }

  /** Helper class to encapsulate column info */
  case class ColumnMetadata(index: Int, structField: StructField)

  /** Helper class that encapsulate an [[RowIndexFilterType]]. */
  case class DeletionVectorDescriptorWithFilterType(
      descriptor: DeletionVectorDescriptor,
      filterType: RowIndexFilterType) {
  }
}

File: kernel-spark/src/main/java/io/delta/kernel/spark/read/DeltaParquetFileFormat.java
/*
 * Copyright (2025) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.delta.kernel.spark.read;

import io.delta.kernel.engine.Engine;
import io.delta.kernel.internal.actions.DeletionVectorDescriptor;
import io.delta.kernel.internal.actions.Metadata;
import io.delta.kernel.internal.actions.Protocol;
import io.delta.kernel.internal.deletionvectors.RoaringBitmapArray;
import io.delta.kernel.internal.util.ColumnMapping;
import io.delta.kernel.internal.util.Tuple2;
import io.delta.kernel.spark.utils.SchemaUtils;
import io.delta.kernel.spark.utils.SerializableKernelRowWrapper;
import io.delta.kernel.types.StructType;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;
import org.apache.hadoop.conf.Configuration;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.catalyst.InternalRow;
import org.apache.spark.sql.execution.datasources.PartitionedFile;
import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat;
import org.apache.spark.sql.sources.Filter;
import org.apache.spark.sql.vectorized.ColumnarBatch;
import org.apache.spark.sql.vectorized.ColumnarBatchRow;
import scala.Function1;
import scala.collection.Iterator;

/**
 * A Delta-aware Parquet file format that supports: - Column Mapping (ID and Name modes) - Deletion
 * Vectors
 *
 * <p>This format wraps Spark's ParquetFileFormat and adds Delta-specific transformations using
 * Kernel APIs.
 */
public class DeltaParquetFileFormat extends ParquetFileFormat {

  // Metadata key for deletion vector descriptor
  public static final String DV_DESCRIPTOR_KEY = "__delta_dv_descriptor";

  // Serializable wrappers for Protocol and Metadata
  private final SerializableKernelRowWrapper protocolWrapper;
  private final SerializableKernelRowWrapper metadataWrapper;
  private final String tablePath;
  private final ColumnMapping.ColumnMappingMode columnMappingMode;

  public DeltaParquetFileFormat(
      Engine kernelEngine, Protocol protocol, Metadata metadata, String tablePath) {
    Objects.requireNonNull(kernelEngine, "kernelEngine is null");
    Objects.requireNonNull(protocol, "protocol is null");
    Objects.requireNonNull(metadata, "metadata is null");
    this.tablePath = Objects.requireNonNull(tablePath, "tablePath is null");

    // Wrap Protocol and Metadata in serializable wrappers
    this.protocolWrapper = new SerializableKernelRowWrapper(protocol.toRow());
    this.metadataWrapper = new SerializableKernelRowWrapper(metadata.toRow());
    this.columnMappingMode = ColumnMapping.getColumnMappingMode(metadata.getConfiguration());
  }

  /** Get Metadata from wrapper */
  private Metadata getMetadata() {
    return Metadata.fromRow(metadataWrapper.getRow());
  }

  public Function1<PartitionedFile, Iterator<InternalRow>> buildReaderWithPartitionValues(
      SparkSession sparkSession,
      org.apache.spark.sql.types.StructType dataSchema,
      org.apache.spark.sql.types.StructType partitionSchema,
      org.apache.spark.sql.types.StructType requiredSchema,
      scala.collection.Seq<Filter> filters,
      scala.collection.immutable.Map<String, String> options,
      Configuration hadoopConf) {

    // Step 1: Convert logical schema to physical schema using Kernel API
    org.apache.spark.sql.types.StructType physicalDataSchema = convertToPhysicalSchema(dataSchema);
    org.apache.spark.sql.types.StructType physicalPartitionSchema =
        convertToPhysicalSchema(partitionSchema);
    org.apache.spark.sql.types.StructType physicalRequiredSchema =
        convertToPhysicalSchema(requiredSchema);

    // Step 2: Translate filters to use physical column names
    scala.collection.immutable.Seq<Filter> physicalFilters =
        columnMappingMode == ColumnMapping.ColumnMappingMode.NONE
            ? convertFiltersToImmutable(filters)
            : io.delta.kernel.spark.utils.ExpressionUtils.convertFiltersToPhysicalNames(
                filters, dataSchema, physicalDataSchema);

    // Step 3: Build standard Parquet reader with physical schema
    Function1<PartitionedFile, Iterator<InternalRow>> baseReader =
        super.buildReaderWithPartitionValues(
            sparkSession,
            physicalDataSchema,
            physicalPartitionSchema,
            physicalRequiredSchema,
            physicalFilters,
            options,
            hadoopConf);

    // Step 4: Wrap reader to apply deletion vector filtering
    return (PartitionedFile file) -> {
      Iterator<InternalRow> baseIterator = baseReader.apply(file);
      return applyDeletionVectorIfNeeded(file, baseIterator, hadoopConf);
    };
  }

  /** Convert logical Spark schema to physical schema using Kernel's ColumnMapping utilities. */
  private org.apache.spark.sql.types.StructType convertToPhysicalSchema(
      org.apache.spark.sql.types.StructType logicalSchema) {
    if (columnMappingMode == ColumnMapping.ColumnMappingMode.NONE) {
      return logicalSchema;
    }

    Metadata metadata = getMetadata();

    // Convert Spark StructType to Kernel StructType
    StructType kernelLogicalSchema = SchemaUtils.convertSparkSchemaToKernelSchema(logicalSchema);
    StructType kernelFullSchema = metadata.getSchema();

    // Use Kernel API to convert to physical schema
    StructType kernelPhysicalSchema =
        ColumnMapping.convertToPhysicalSchema(
            kernelLogicalSchema, kernelFullSchema, columnMappingMode);

    // Convert back to Spark StructType
    return SchemaUtils.convertKernelSchemaToSparkSchema(kernelPhysicalSchema);
  }

  /** Convert Seq to immutable Seq for compatibility. */
  private scala.collection.immutable.Seq<Filter> convertFiltersToImmutable(
      scala.collection.Seq<Filter> filters) {
    if (filters instanceof scala.collection.immutable.Seq) {
      return (scala.collection.immutable.Seq<Filter>) filters;
    }
    return scala.collection.JavaConverters.asScalaBuffer(
            scala.collection.JavaConverters.seqAsJavaList(filters))
        .toSeq();
  }

  /**
   * Apply deletion vector filtering if present. Supports both vectorized (ColumnarBatch) and
   * non-vectorized (InternalRow) data from Parquet reader.
   */
  @SuppressWarnings("unchecked")
  private Iterator<InternalRow> applyDeletionVectorIfNeeded(
      PartitionedFile file, Iterator<InternalRow> dataIterator, Configuration hadoopConf) {

    Optional<DeletionVectorDescriptor> dvDescriptorOpt = extractDeletionVectorDescriptor(file);

    if (!dvDescriptorOpt.isPresent()) {
      return dataIterator;
    }

    // Load deletion vector using Kernel API
    RoaringBitmapArray deletionVector = loadDeletionVector(dvDescriptorOpt.get(), hadoopConf);

    // Filter out deleted rows - handle both vectorized and row-based data
    // Cast to Iterator<Object> since Parquet may return ColumnarBatch or InternalRow
    Iterator<Object> objectIterator = (Iterator<Object>) (Iterator<?>) dataIterator;
    return new DeletionVectorFilterIterator(objectIterator, deletionVector);
  }

  /** Extract deletion vector descriptor from PartitionedFile metadata. */
  private Optional<DeletionVectorDescriptor> extractDeletionVectorDescriptor(PartitionedFile file) {
    scala.collection.immutable.Map<String, Object> metadata =
        file.otherConstantMetadataColumnValues();

    scala.Option<Object> dvOption = metadata.get(DV_DESCRIPTOR_KEY);
    if (dvOption.isDefined()) {
      Object dvObj = dvOption.get();
      if (dvObj instanceof DeletionVectorDescriptor) {
        return Optional.of((DeletionVectorDescriptor) dvObj);
      }
    }
    return Optional.empty();
  }

  /** Load deletion vector bitmap using Kernel API. */
  private RoaringBitmapArray loadDeletionVector(
      DeletionVectorDescriptor dvDescriptor, Configuration hadoopConf) {
    try {
      // Create a new engine for this task
      Engine engine = io.delta.kernel.defaults.engine.DefaultEngine.create(hadoopConf);
      Tuple2<DeletionVectorDescriptor, RoaringBitmapArray> result =
          io.delta.kernel.internal.deletionvectors.DeletionVectorUtils.loadNewDvAndBitmap(
              engine, tablePath, dvDescriptor);
      return result._2;
    } catch (Exception e) {
      throw new RuntimeException("Failed to load deletion vector", e);
    }
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) return true;
    if (!(other instanceof DeltaParquetFileFormat)) return false;

    DeltaParquetFileFormat that = (DeltaParquetFileFormat) other;
    return Objects.equals(this.tablePath, that.tablePath)
        && Objects.equals(this.columnMappingMode, that.columnMappingMode);
  }

  @Override
  public int hashCode() {
    return Objects.hash(tablePath, columnMappingMode);
  }

  /**
   * Iterator that filters out rows marked as deleted in the deletion vector. Supports both
   * vectorized (ColumnarBatch) and non-vectorized (InternalRow) data.
   */
  private static class DeletionVectorFilterIterator
      extends scala.collection.AbstractIterator<InternalRow> {
    private final Iterator<Object> underlying;
    private final RoaringBitmapArray deletionVector;
    private long currentRowIndex = 0;

    // For handling ColumnarBatch - use Scala Iterator
    private scala.collection.Iterator<InternalRow> currentBatchIterator = null;

    // Type handlers map for processing different data formats
    private final Map<Class<?>, Function<Object, InternalRow>> typeHandlers;

    DeletionVectorFilterIterator(Iterator<Object> underlying, RoaringBitmapArray deletionVector) {
      this.underlying = underlying;
      this.deletionVector = deletionVector;

      // Initialize type handlers
      this.typeHandlers = new HashMap<>();
      typeHandlers.put(ColumnarBatch.class, this::handleColumnarBatch);
      typeHandlers.put(ColumnarBatchRow.class, this::handleColumnarBatchRow);
      typeHandlers.put(InternalRow.class, this::handleInternalRow);
    }

    @Override
    public boolean hasNext() {
      // First check if we have rows from current batch
      if (currentBatchIterator != null && currentBatchIterator.hasNext()) {
        return true;
      }

      // Try to get next batch or row
      return underlying.hasNext();
    }

    @Override
    public InternalRow next() {
      if (!hasNext()) {
        throw new NoSuchElementException();
      }

      // If we have rows from current batch, return next one
      if (currentBatchIterator != null && currentBatchIterator.hasNext()) {
        return currentBatchIterator.next();
      }

      // Get next item from underlying iterator
      Object next = underlying.next();

      // Use type handlers map to process different data formats
      Function<Object, InternalRow> handler = typeHandlers.get(next.getClass());
      if (handler != null) {
        return handler.apply(next);
      } else {
        throw new RuntimeException(
            "Unexpected row type from Parquet reader: " + next.getClass().getName());
      }
    }

    /** Handle vectorized ColumnarBatch data */
    private InternalRow handleColumnarBatch(Object obj) {
      ColumnarBatch batch = (ColumnarBatch) obj;
      List<InternalRow> filteredRows = filterColumnarBatch(batch);
      // Convert Java Iterator to Scala Iterator
      currentBatchIterator =
          scala.collection.JavaConverters.asScalaIterator(filteredRows.iterator());
      return currentBatchIterator.next();
    }

    /**
     * Handle ColumnarBatchRow - vectorized reader enabled but returns immutable rows. This is not
     * efficient and should only affect wide tables.
     */
    private InternalRow handleColumnarBatchRow(Object obj) {
      ColumnarBatchRow columnarRow = (ColumnarBatchRow) obj;
      // Need to copy the row since ColumnarBatchRow is immutable
      InternalRow row = columnarRow.copy();
      // Filter out deleted rows
      while (deletionVector.contains(currentRowIndex)) {
        currentRowIndex++;
        if (!underlying.hasNext()) {
          throw new NoSuchElementException();
        }
        Object next = underlying.next();
        if (next instanceof ColumnarBatchRow) {
          row = ((ColumnarBatchRow) next).copy();
        } else {
          row = (InternalRow) next;
        }
      }
      currentRowIndex++;
      return row;
    }

    /** Handle non-vectorized InternalRow data */
    private InternalRow handleInternalRow(Object obj) {
      InternalRow row = (InternalRow) obj;
      // Filter out deleted rows
      while (deletionVector.contains(currentRowIndex)) {
        currentRowIndex++;
        if (!underlying.hasNext()) {
          throw new NoSuchElementException();
        }
        row = (InternalRow) underlying.next();
      }
      currentRowIndex++;
      return row;
    }

    /** Filter ColumnarBatch by deletion vector. Returns list of non-deleted rows. */
    private List<InternalRow> filterColumnarBatch(ColumnarBatch batch) {
      List<InternalRow> result = new ArrayList<>();
      int numRows = batch.numRows();

      for (int i = 0; i < numRows; i++) {
        if (!deletionVector.contains(currentRowIndex + i)) {
          result.add(batch.getRow(i).copy());
        }
      }

      currentRowIndex += numRows;
      return result;
    }
  }
}
